Playwright: Update the I18N TeamCity build for Playwright

dpasque opened the pull request:
#### Changes proposed in this Pull Request

With PRs #57707 and #56940, our I18N tests are now in Playwright!

As a result, we need to update the TeamCity build job to use Playwright and Jest.

Otherwise, this should be an exact port of the old Selenium build!

Key changes:
- use TeamCity to define a list of locales to be tested, under `LOCALES` in the build parameter.
- new function to return the comma-delimited list of locales as an array of strings in `environment.ts`.
- refactor all three i18n tests to use `describe.each` or `it.each` to iterate over the list of locales.
- changes to block testing steps in Editor: i18n to ensure more reliable tests.

Details:
Previously the i18n tests ran on a loop that was baked into the TeamCity configuration, which looked like this:
```
for locale in ${'$'}{LOCALE[@]}; do
    """.trimIndent()
	  BROWSERLOCALE="${'$'}{locale}" yarn magellan --config=magellan-i18n.json --max_workers=%E2E_WORKERS% --local_browser=chrome --mocha_args="--reporter mocha-multi-reporters --reporter-options configFile=mocha-reporter.json" || true
done
```

This causes issues with TeamCity's parsing of the build log and all sorts of other intelligent things TeamCity does for each build. With this change, instead of using a baked-in for loop, the locales that must be tested will be defined as an comma-delimited environment variable. This variable is parsed by the `TesEnvironment.LOCALES` method which returns an array of strings representing the locale.

Each i18n test has been fitted with Jest's native parametrization which will result in nice output that's friendly for TeamCity to parse, as well as eliminating repetitive work.

Additionally, `I18N: Editor` spec was refactored to remove unnecessary logic in order to make it easier to follow.

Sample output:
```
 I18N: Gutenboarding
16:27:25       Locale: en
16:27:25         ✓ Navigate to /new (474 ms)
16:27:25         ✓ Change the locale (en) (669 ms)
16:27:25         ✓ Display translations on aquire intent page (en) (407 ms)
16:27:25         ✓ Display translations on design selector page (en) (346 ms)
16:27:25         ✓ Display translations on style preview page (en) (611 ms)
16:27:25         ✓ Display translations on domains page (en) (960 ms)
16:27:25         ✓ Display translations on features page (en) (413 ms)
16:27:25         ✓ Display translations on plans page (en) (337 ms)
16:27:25       Locale: es
16:27:25         ✓ Navigate to /new (171 ms)
16:27:25         ✓ Change the locale (es) (1043 ms)
16:27:25         ✓ Display translations on aquire intent page (es) (325 ms)
16:27:25         ✓ Display translations on design selector page (es) (414 ms)
16:27:25         ✓ Display translations on style preview page (es) (554 ms)
16:27:25         ✓ Display translations on domains page (es) (1097 ms)
16:27:25         ✓ Display translations on features page (es) (474 ms)
16:27:25         ✓ Display translations on plans page (es) (467 ms)
```

#### Testing instructions

- [x] I18N Tests
  - [x] TeamCity
  - [ ] Local

Closes #57777

Commit:
Update the I18N TeamCity build for Playwright

Update ENV variable name

Fix environment prep

scinos commented:
> Unfortunately, with these we have to merge them before we test them! 

If it's a change in a specific step, couldn't you just run the build for this branch? Am I missing something?

I've run `I18N Tests` build for this branch, and it "works" in the sense that the script is right... but each iteration of the loop finds zero tests to run. Extract from the logs:

```
05:03:14   + for locale in ${LOCALE[@]}
05:03:14   + BROWSERLOCALE=es
05:03:14   + yarn jest --reporters=jest-teamcity --reporters=default --maxWorkers=16 --group=i18n
05:03:17   Test Suites: 0 of 56 total
05:03:17   Tests:       0 total
05:03:17   Snapshots:   0 total
05:03:17   Time:        0.241 s
05:03:17   Ran all test suites.
```

yuliyan commented on .teamcity/_self/projects/WPComTests.kt:
```diff
@@ -742,47 +742,46 @@ private object I18NTests : BuildType({
+				# Run the test
+				IFS="," read -r -a LOCALE <<< "%LOCALES%"
+				for locale in ${'$'}{LOCALE[@]}; do
+					BROWSERLOCALE="${'$'}{locale}" yarn jest --reporters=jest-teamcity --reporters=default --maxWorkers=%E2E_WORKERS% --group=i18n
```
I think the env variable should `LOCALE` instead of `BROWSERLOCALE` - https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/browser-helper.ts#L36. I think that's likely the reason why @scinos got [zero tests running on his testing](https://github.com/Automattic/wp-calypso/pull/57934#issuecomment-965970094).

dpasque replied:
Ack! Nice catch, thank you! I'll push that up!

dpasque commented:
>If it's a change in a specific step, couldn't you just run the build for this branch? Am I missing something?

Nope, but clearly I am! 😆  Thanks for the heads up, I don't think I realized this worked like this, that's really helpful!

Is it only brand new steps that have to merge into trunk first before they're executable?

dpasque commented:
>but each iteration of the loop finds zero tests to run. 

I believe this is because the PRs that add the Playwright I18N tests haven't merged quite yet!

dpasque commented:
I18N tests are now running correctly! It looks like there's some spec failures, but the TeamCity test execution pieces seem correct & in place

dpasque commented:
@scinos Checking in on this one - does this now look good to go? Thanks! 😄

scinos commented:
Looks like all tests are in English?

![image](https://user-images.githubusercontent.com/975703/141812319-cebbb92b-040b-4007-8f2c-30823946f034.png)

I'm slightly worried about running `jest` in a loop. I'm not 100% sure TeamCity will pick all the tests results, maybe it only picks the first (or last?) iteration of the loop.

worldomonation commented:
> I'm slightly worried about running jest in a loop. I'm not 100% sure TeamCity will pick all the tests results, maybe it only picks the first (or last?) iteration of the loop.

Personally I think this is the ideal use case of `test.each` - or parametrization in general. 

If the locale being read then set from an environment variable is not an inherent part of Calypso that needs to be tested, then I'd much rather see parametrized test be used instead.

dpasque commented:
>If the locale being read then set from an environment variable is not an inherent part of Calypso that needs to be tested, then I'd much rather see parametrized test be used instead.

That works! I will update this to take in a comma delimited list of locales, and then we can add some library methods to split that and loop over the tests under test.

@yuliyan -- want me to add those test file updates in this PR right here so they are in sync with the TeamCity change? I didn't know if that would be disruptive to work you have in flight!

yuliyan commented:
@dpasque, yes, I'd appreciate it!

worldomonation commented:
@dpasque If you have a list of things that need done, I can take over this from you?

dpasque commented:
>@dpasque If you have a list of things that need done, I can take over this from you?

That might be really helpful so I can get the other small things all wrapped up, thank you so much!

Here's what's left for this PR:

- [ ] Change the bash script in the TC build to pass in the comma delimited list of locales and skip the loop
- [ ] Make a library helper to parse that into an array, and maybe also a jest helper that runs a parent describe block for that array
- [ ] Update the existing I18N specs to use that

worldomonation committed:
Revamp how the parametrization is handled for i18n tests.
- define new constant in `environment.ts` to obtain a list of locales from environment variable.
- refactor `wp-i18n__editor.ts` spec to accept a list of locales to parametrize on.
- adjust guards to ensure only locales with defined translations run the test.

worldomonation commented:
Pushed a new revision that:
- reads the TeamCity env var (comma-delimited) and returns it as an array.
- revamp the `wp-i18n__editor` spec to ensure only languages with translations defined run the test.
- adjust suite naming schema.
- replace usage of `async () =>` with `async function` which is the correct usage as per style guide.
- manually override timeouts when inserting then verifying translations on the block to 10000ms.

scinos commented on .teamcity/_self/projects/WPComTests.kt:
```diff
@@ -730,59 +730,50 @@ private object I18NTests : BuildType({
-				""".trimIndent()
-			dockerRunParameters = "-u %env.UID% --security-opt seccomp=.teamcity/docker-seccomp.json --shm-size=8gb"
+				# Run the test
+				IFS="," read -r -a TARGET_LOCALES <<< "%LOCALES%"
+				for locale in ${'$'}{TARGET_LOCALES[@]}; do
+					LOCALE="${'$'}{locale}" yarn jest --reporters=jest-teamcity --reporters=default --maxWorkers=%E2E_WORKERS% --group=i18n
```
Now that we are using `describe.each( locales )` I guess we don't need this loop, right?

worldomonation replied:
That is correct, though I should mark this PR as draft as I am still working on changes.

Thanks for reviewing so quickly though!

scinos commented on packages/calypso-e2e/src/environment.ts:
```diff
@@ -1 +1,11 @@
+ * @returns {string[]} Array of lowercase strings.
+ */
+export const LOCALES = (): string[] => {
+	const locales = process.env.I18N_LOCALES ?? '';
```
Where is `I18N_LOCALES` set?

worldomonation replied:
I've revised it to use LOCALES in a following revision (that I haven't pushed yet). Didn't know the full context of this change so didn't realize that LOCALES was set as a TC param.

worldomonation commented:
Setting as draft as this is still a WIP.

worldomonation committed:
Additional fixes:
- update WPCOM TeamCity build config.
- fix environment variable
- replace beforeAll with step in Editor spec.

worldomonation committed:
Update 'Render block title translations' step to work more reliably.

worldomonation commented on .teamcity/_self/projects/WPComTests.kt:
```diff
@@ -730,59 +730,48 @@ private object I18NTests : BuildType({
-				""".trimIndent()
-			dockerRunParameters = "-u %env.UID% --security-opt seccomp=.teamcity/docker-seccomp.json --shm-size=8gb"
+				# Run the test
+				yarn jest --reporters=jest-teamcity --reporters=default --maxWorkers=%E2E_WORKERS% --group=i18n
```
Note that we are no longer using a for loop here, as the looping is now done using Jest's `describe.each`.

worldomonation commented on .teamcity/_self/projects/WPComTests.kt:
```diff
@@ -730,59 +730,48 @@ private object I18NTests : BuildType({
+				export DEBUG=pw:api
+				export HEADLESS=true
+				export TARGET_DEVICE=desktop
+				export LOCALES=%LOCALES%
```
I learned that this translates the build parameter to an environment variable.

worldomonation commented on test/e2e/specs/specs-i18n/wp-i18n__editor.ts:
```diff
@@ -214,85 +214,101 @@ const translations: Translations = {
-	() => {
-		let gutenbergEditorPage: GutenbergEditorPage;
-		let page: Page;
+describe( 'I18N: Editor', function () {
```
I've renamed the suites to be more consistent. They will now all start with an 'I18N', followed by a colon and name of the major feature being tested.

worldomonation commented on test/e2e/specs/specs-i18n/wp-i18n__editor.ts:
```diff
@@ -214,85 +214,101 @@ const translations: Translations = {
+describe( 'I18N: Editor', function () {
+	let page: Page;
+	// Filter out the locales that do not have valid translation content defined above.
+	const locales = Object.keys( translations ).filter( ( locale ) =>
```
This portion is important in this spec file.

We define a long list of locales that need to be tested in TeamCity, but for this spec file, only `fr`, `he` and `en` have translations defined. As such, we cannot use the raw array of locales.

The solution is to filter out locales other than the three with translations.

As an added benefit, this removes the extraneous `describeSkipNoTranslations` from Line 265 on the left hand file and thus one layer of unnecessary nesting.

worldomonation commented on test/e2e/specs/specs-i18n/wp-i18n__editor.ts:
```diff
@@ -214,85 +214,101 @@ const translations: Translations = {
 
-			const changeUILanguageFlow = new ChangeUILanguageFlow( page );
-			await changeUILanguageFlow.changeUILanguage( locale );
+	describe.each( locales )( `Locale: %s`, function ( locale ) {
```
Parametrize the setup, start new post and block verification for each language that has translation defined.

worldomonation commented on test/e2e/specs/specs-i18n/wp-i18n__editor.ts:
```diff
@@ -214,85 +214,101 @@ const translations: Translations = {
 
-				beforeAll( async () => {
-					frame = await gutenbergEditorPage.getEditorFrame();
+				const blockTimeout = 10 * 1000;
```
Manual timeout override.

worldomonation commented on test/e2e/specs/specs-i18n/wp-i18n__editor.ts:
```diff
@@ -214,85 +214,101 @@ const translations: Translations = {
-					] );
 
+					// Ensure the block is highlighted.
+					await frame.waitForSelector(
+						`:is( ${ block.blockEditorSelector }.is-selected, ${ block.blockEditorSelector }.has-child-selected)`,
+						{ timeout: blockTimeout }
```
Instead of using `Promise.race`, this now uses the [`:is`](https://playwright.dev/docs/selectors#selecting-elements-matching-one-of-the-conditions) pseudo-CSS selector. This selector will match on a list of selectors passed in as argument.

This is necessary as in some cases (eg. Contact Form block) there exists sub-blocks, and if these child blocks are selected then the attributed added is `has-child-selected` instead of `is-selected`.

worldomonation commented on test/e2e/specs/specs-i18n/wp-i18n__editor.ts:
```diff
@@ -214,85 +214,101 @@ const translations: Translations = {
+					// If on block insertion, one of the sub-blocks are selected, click on
+					// the first button in the floating toolbar which selects the overall
+					// block.
+					if ( await frame.isVisible( '.block-editor-block-parent-selector__button' ) ) {
+						await frame.click( '.block-editor-block-parent-selector__button' );
+					}
```
Only require clicking on the floating toolbar to select the parent block if the block currently selected is a child block.

worldomonation commented on test/e2e/specs/specs-i18n/wp-i18n__gutenboarding.ts:
```diff
@@ -5,62 +5,62 @@
-		await page.goto( DataHelper.getCalypsoURL( 'new' ) );
-		gutenboardingFlow = new GutenboardingFlow( page );
-	} );
+	describe.each( TestEnvironment.LOCALES() )( 'Locale: %s', function ( locale ) {
```
This spec supports every locale under test defined in TeamCity and thus is a straight loop over all of them.

worldomonation commented on test/e2e/specs/specs-i18n/wp-i18n__logged-out-redirect.ts:
```diff
@@ -2,19 +2,17 @@
-const locale = BrowserHelper.getLocale();
-
-describe( `Logged out homepage redirect test @i18n (${ locale })`, function () {
+describe.skip( 'I18N: Homepage Redirect', function () {
```
I've skipped this test for now because this test doesn't actually do anything.

What I observe is as follows:
1. browser is launched
2. page navigates to `https://wordpress.com`

Then nothing occurs after that, until the next call in Line 20 times out since nothing happened.

@yuliyan, is there something missing from this portion? I am not sure how the original Selenium spec looked like, so perhaps it's missing a portion like setting the locale.

yuliyan replied:
It's testing the locale redirects for https://wordpress.com, that are based on the browser language settings that set the Accept-Language header for request. For example, if you set Spanish as your primary browser language and you go to `https://wordpress.com` (logged-out), you should be redirected to `https://wordpress.com/es`.
For Playwright, we specify [the `locale` config when creating new browser context](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/browser-helper.ts#L89) in order to set the browser language.

worldomonation replied:
> It's testing the locale redirects for https://wordpress.com, that are based on the browser language settings that set the Accept-Language header for request. For example, if you set Spanish as your primary browser language and you go to `https://wordpress.com` (logged-out), you should be redirected to `https://wordpress.com/es`. For Playwright, we specify [the `locale` config when creating new browser context](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/browser-helper.ts#L89) in order to set the browser language.

Got it, thanks. I see why this test does not seem to do 'anything' at this moment - the BrowserContext is created with `en` as the locale and this persists for the duration of the test and this is why the redirect expectation fails.

I have some ideas about how to work around this, by implementing a new method for BrowserManager to help set the locale. I'll give that a try and if it works incorporate them to the PR.

worldomonation commented:
This PR is ready for review. I have taken over from Dan.

worldomonation committed:
Changes to how the locale is set for new browsercontexts:
- optional parameter in LaunchOptions interface for locale and logger.
- in the spec, launch new browser context per custom locale.

worldomonation committed:
Use trace-enabled artifact rule

worldomonation committed:
Update test step name of Homepage Redirect

worldomonation commented:
All tests green and running. Time to merge!
